File: android/app/src/main/java/dev/terminaldeck/android/ui/KeyBar.kt
package dev.terminaldeck.android.ui

/**
 * The touch key row, and the sticky Ctrl behind it.
 *
 * Transcribed from `pwa/src/keybar.ts`, including the reasoning, because the two clients should
 * not disagree about what Ctrl does.
 *
 * ## Why this exists at all
 *
 * The Android software keyboard has no Esc, no Tab, no Ctrl and no arrows, and it hides `|`, `/`,
 * `-` and `~` behind two page flips. Every one of those is on the critical path of using a
 * terminal: Ctrl+C stops a runaway process, Tab completes a path, the up arrow recalls the last
 * command, `~` starts a home path and `|` builds a pipeline. Without this row the app renders a
 * terminal that can type prose at a shell and nothing else.
 *
 * ## Why Ctrl is sticky rather than held
 *
 * A finger cannot hold a chord. Ctrl therefore arms, and the next key folds into it — the same
 * interaction every phone keyboard uses for shift. Arming is a toggle, so a mistaken tap is undone
 * by tapping again rather than by being forced to spend it on a key.
 *
 * An armed Ctrl is always spent by the next key, **including a key it cannot combine with**.
 * Leaving it armed after a key it does not apply to means it fires on the key after that, which is
 * how a sticky modifier turns a `w` into a Ctrl+W and closes something.
 *
 * ## No Compose in this file
 *
 * The state machine is where the bugs live, and it is unit-tested on a plain JVM. The row that
 * draws it is in `TerminalScreen`.
 */

enum class KeyBarKey(val id: String, val label: String, val title: String) {
    Esc("esc", "esc", "Escape"),
    Tab("tab", "tab", "Tab"),
    Ctrl("ctrl", "ctrl", "Control — applies to the next key"),
    Up("up", "↑", "Up arrow"),
    Down("down", "↓", "Down arrow"),
    Left("left", "←", "Left arrow"),
    Right("right", "→", "Right arrow"),
    Pipe("pipe", "|", "Pipe"),
    Slash("slash", "/", "Slash"),
    Dash("dash", "-", "Hyphen"),
    Tilde("tilde", "~", "Tilde"),
}

/**
 * The fixed bar, in tap order, and it is only five keys now.
 *
 * This used to be every key in the enum in one horizontal scroll, which is the shape iOS threw out
 * and wrote down why in `KeyPlan`: a scrolling bar has no fixed positions, so no muscle memory ever
 * forms and the key you want is always mid-swipe. The bar now carries only what is pressed *while
 * typing a command* — `esc` `tab` `ctrl` `↑` `↓` — and everything else (the horizontal arrows, the
 * symbols, the signals, alt and the function keys) moved into the grid the *more keys* button opens.
 * See [KeyGrid.groups].
 *
 * Ctrl sits third rather than first: on a phone held one-handed the far left of the row is the
 * least reachable spot, and Esc and Tab are the two that get hit blind. The enum keeps all of its
 * members — [KeyBar.press] and its tests still speak for the arrows and symbols the grid now draws —
 * but the bar shows this handful.
 */
val KEY_BAR: List<KeyBarKey> = listOf(
    KeyBarKey.Esc,
    KeyBarKey.Tab,
    KeyBarKey.Ctrl,
    KeyBarKey.Up,
    KeyBarKey.Down,
)

/** What a press produced: bytes for the session, and the modifier state after it. */
data class KeyPress(val data: String, val ctrl: Boolean)

object KeyBar {

    /**
     * The ASCII control byte for a character, or null when there is not one.
     *
     * This is the actual rule rather than a table of guesses: a control character is the printable
     * one with bits 6 and 7 cleared, which is defined for `@` through `_` and, by the same masking,
     * the lowercase letters. Space is `@` under another name and `?` is the odd one at 0x7F.
     *
     * Everything else returns null, deliberately. Terminals disagree about Ctrl+`/` — the mask says
     * 0x0F, several emulators send 0x1F — and this client cannot ask which one is on the other end
     * of the socket, so it does not invent an answer.
     */
    fun controlByteFor(char: Char): String? = when {
        char == ' ' -> "\u0000"
        char == '?' -> "\u007f"
        char.code in 0x40..0x5f -> (char.code and 0x1f).toChar().toString()
        char.code in 0x61..0x7a -> (char.code and 0x1f).toChar().toString()
        else -> null
    }

    /**
     * Press a key from the row.
     *
     * @param ctrl whether Ctrl is currently armed.
     */
    fun press(key: KeyBarKey, ctrl: Boolean): KeyPress = when (key) {
        KeyBarKey.Ctrl -> KeyPress("", !ctrl)

        // Arrows under Ctrl are the word-motion sequences xterm defines; sending the plain arrow
        // instead would silently drop the modifier the user just armed.
        KeyBarKey.Up -> KeyPress(if (ctrl) CSI + "1;5A" else CSI + "A", false)
        KeyBarKey.Down -> KeyPress(if (ctrl) CSI + "1;5B" else CSI + "B", false)
        KeyBarKey.Right -> KeyPress(if (ctrl) CSI + "1;5C" else CSI + "C", false)
        KeyBarKey.Left -> KeyPress(if (ctrl) CSI + "1;5D" else CSI + "D", false)

        KeyBarKey.Esc -> KeyPress(ESC, false)
        KeyBarKey.Tab -> KeyPress(if (ctrl) CSI + "Z" else "\t", false)

        // A character key: fold Ctrl in if there is a control byte for it, and spend the modifier
        // either way.
        KeyBarKey.Pipe -> character('|', ctrl)
        KeyBarKey.Slash -> character('/', ctrl)
        KeyBarKey.Dash -> character('-', ctrl)
        KeyBarKey.Tilde -> character('~', ctrl)
    }

    /**
     * Escape, written as an escape.
     *
     * A raw 0x1B byte in source is invisible in every diff and in every editor, and a sequence
     * missing its introducer types `[A` at the shell instead of moving the cursor — which looks
     * like a broken key rather than a broken constant.
     */
    private const val ESC = "\u001b"

    /** Escape followed by `[`, spelled out so no sequence below hides its introducer. */
    private const val CSI = "\u001b["

    private fun character(char: Char, ctrl: Boolean): KeyPress =
        KeyPress(if (ctrl) controlByteFor(char) ?: char.toString() else char.toString(), false)
}

/**
 * What is in the grid the *more keys* button opens, and what each key sends.
 *
 * Transcribed from `ios/TerminalDeck/Terminal/KeyPlan.swift`, groups and order and all, so the two
 * phones do not disagree about what a key does or where a thumb finds it. Pure data, like the rest
 * of this file, so the one thing worth getting wrong - does every key send the bytes a terminal
 * acts on - is answerable without a screenshot. The Compose that draws it is [KeyGridSheet].
 *
 * ## Why a second model beside [KeyBarKey]
 *
 * The bar's engine ([KeyBar.press]) is a tested state machine about one thing - the sticky Ctrl -
 * and its keys all resolve to a chord or a character. The grid holds kinds the bar never has: an
 * app action (copy, paste), a raw signal byte, a second sticky modifier (alt). Bending the enum to
 * carry all of that would break the very tests that keep Ctrl honest, so the grid gets its own flat
 * data and shares nothing but the escape bytes below.
 *
 * ## Cursor keys send the CSI form, matching the bar rather than the terminal's mode
 *
 * iOS asks the emulator whether it is in application-cursor mode and sends `ESC O A` when it is.
 * This client's bar has always sent the plain `ESC [ A` regardless - see [KeyBar.press] - so the
 * grid does the same: a grid whose arrows disagreed with the bar's directly above it would be the
 * worse bug. Lifting both to honour the mode is a real change and belongs to the bar's engine, not
 * to the surface that draws the keys.
 *
 * The control bytes are built from their code points ([Char.toChar] of 3, 27, ...) rather than
 * written as escapes: a raw 0x1B in source is invisible in every diff and every editor, and the
 * number says plainly which byte it is.
 */
enum class GridModifier { Ctrl, Meta }

/** What pressing a grid key does. The three that are not bytes are handled by the screen. */
sealed interface GridAction {
    /** Literal bytes for the session, already encoded. */
    data class Bytes(val data: String) : GridAction
    /** A cursor key - resolved to `ESC [ <final>` at press time. See the header on the mode. */
    data class Cursor(val final: Char) : GridAction
    /** A sticky modifier the terminal owns; [GridModifier.Meta] is alt, [GridModifier.Ctrl] the bar's. */
    data class Mod(val modifier: GridModifier) : GridAction
    /** Copy the selection or the screen - the app's business, not the wire's. */
    data object Copy : GridAction
    /** Paste the clipboard. */
    data object Paste : GridAction
}

/**
 * One key cap in the grid.
 *
 * [label] is what is written on it - a glyph for the arrows, a word for the rest, because `esc` and
 * `home` have no icon anyone would know. [title] is what a screen reader says and what a test finds
 * the cap by when the label is a symbol. [repeats] is true only for the arrows, where holding to
 * move the cursor is wanted; holding `~` for forty tildes, or `^C` for forty interrupts, is not.
 */
data class GridKey(
    val label: String,
    val title: String,
    val action: GridAction,
    val repeats: Boolean = false,
)

/** A labelled run of keys - the header is the whole point, against a wall of identical squares. */
data class GridGroup(val title: String, val keys: List<GridKey>)

object KeyGrid {

    /** Escape (0x1B), built from its code point so no raw control byte hides in the source. */
    private val ESC: String = 27.toChar().toString()

    /** Escape followed by `[`, spelled out so no sequence below loses its introducer. */
    private val CSI: String = ESC + "["

    /** `ESC [ <final>`, the CSI cursor/edit form. Home and End travel this way too - they are cursor keys. */
    fun cursorBytes(final: Char): String = CSI + final

    /**
     * The bytes for a function key, as xterm sends them - from a table, not a formula.
     *
     * F1-F4 are SS3 (`ESC O P`...`ESC O S`); F5 up are CSI with a number that skips 16, 22 and 27,
     * because the DEC keyboard those came from had keys a PC one does not. Written out because that
     * is the only way it is right.
     */
    fun functionBytes(number: Int): String = when (number) {
        1 -> ESC + "OP"
        2 -> ESC + "OQ"
        3 -> ESC + "OR"
        4 -> ESC + "OS"
        5 -> CSI + "15~"
        6 -> CSI + "17~"
        7 -> CSI + "18~"
        8 -> CSI + "19~"
        9 -> CSI + "20~"
        10 -> CSI + "21~"
        11 -> CSI + "23~"
        12 -> CSI + "24~"
        else -> ""
    }

    private fun symbolTitle(char: String): String = when (char) {
        "|" -> "Pipe"
        "/" -> "Slash"
        "\\" -> "Backslash"
        "-" -> "Hyphen"
        "_" -> "Underscore"
        "~" -> "Tilde"
        ":" -> "Colon"
        "*" -> "Asterisk"
        else -> char
    }

    /**
     * The grid, in the order the groups are read - by how often a group is wanted, not by taxonomy.
     *
     * Edit and signals are what a thumb reaches for in a hurry and open at the top; symbols are what
     * it reaches for while typing; the function keys nobody presses twice a week are last.
     */
    val groups: List<GridGroup> = listOf(
        GridGroup("Edit", listOf(
            GridKey("copy", "Copy the selection, or the screen", GridAction.Copy),
            GridKey("paste", "Paste", GridAction.Paste),
        )),
        GridGroup("Signals", listOf(
            // Written out rather than composed from Ctrl, because these four are the ones worth one
            // tap: interrupt (^C=3), end-of-file (^D=4), suspend (^Z=26), clear (^L=12).
            GridKey("^C", "Interrupt", GridAction.Bytes(3.toChar().toString())),
            GridKey("^D", "End of file", GridAction.Bytes(4.toChar().toString())),
            GridKey("^Z", "Suspend", GridAction.Bytes(26.toChar().toString())),
            GridKey("^L", "Clear the screen", GridAction.Bytes(12.toChar().toString())),
        )),
        GridGroup("Navigation", listOf(
            GridKey("←", "Left arrow", GridAction.Cursor('D'), repeats = true),
            GridKey("→", "Right arrow", GridAction.Cursor('C'), repeats = true),
            GridKey("home", "Home", GridAction.Cursor('H')),
            GridKey("end", "End", GridAction.Cursor('F')),
            GridKey("pgup", "Page up", GridAction.Bytes(CSI + "5~")),
            GridKey("pgdn", "Page down", GridAction.Bytes(CSI + "6~")),
        )),
        GridGroup("Symbols", listOf("|", "/", "\\", "-", "_", "~", ":", "*").map { char ->
            GridKey(char, symbolTitle(char), GridAction.Bytes(char))
        }),
        GridGroup("Modifiers", listOf(
            GridKey("alt", "Alt - sends the next key with an Escape prefix", GridAction.Mod(GridModifier.Meta)),
        )),
        GridGroup("Function", (1..12).map { number ->
            GridKey("F$number", "Function key $number", GridAction.Bytes(functionBytes(number)))
        }),
    )
}

File: pwa/src/keybar.ts
/**
 * The touch key bar, and the sticky Ctrl behind it.
 *
 * ## Why this exists at all
 *
 * The iOS software keyboard has no Esc, no Tab, no Ctrl and no arrows, and it
 * hides `|`, `/`, `-` and `~` behind two page flips. Every one of those is on
 * the critical path of using a terminal: Ctrl+C stops a runaway process, Tab
 * completes a path, the up arrow recalls the last command, `~` starts a home
 * path and `|` builds a pipeline. Without this row the client renders a
 * terminal that can type prose at a shell and nothing else.
 *
 * ## Why Ctrl is sticky rather than held
 *
 * A finger cannot hold a chord. Ctrl therefore arms, and the next key folds
 * into it — the same interaction every phone keyboard uses for shift. Arming
 * is a toggle, so a mistaken tap is undone by tapping again rather than by
 * being forced to spend it on a key.
 *
 * An armed Ctrl is always spent by the next key, including a key it cannot
 * combine with. Leaving it armed after a key it does not apply to means it
 * fires on the key after that, which is how a sticky modifier turns a `w` into
 * a Ctrl+W and closes something.
 *
 * There is no DOM in this half of the file on purpose: the state machine is
 * where the bugs live, and vitest here runs with no DOM environment.
 */

/* ------------------------------------------------------------------ keys -- */

export type KeyBarKeyId =
  | 'esc'
  | 'tab'
  | 'ctrl'
  | 'up'
  | 'down'
  | 'left'
  | 'right'
  | 'pipe'
  | 'slash'
  | 'dash'
  | 'tilde'

export interface KeyBarKey {
  id: KeyBarKeyId
  /** What the button says. */
  label: string
  /** Read out to screen readers, where the glyph alone is not a word. */
  title: string
}

/**
 * The row, in tap order.
 *
 * Ctrl sits third rather than first: on a phone held one-handed the far left
 * of the row is the least reachable spot, and Esc and Tab are the two that get
 * hit blind. The four punctuation keys are last because they are the ones a
 * user will look down at.
 */
export const KEY_BAR: readonly KeyBarKey[] = [
  { id: 'esc', label: 'esc', title: 'Escape' },
  { id: 'tab', label: 'tab', title: 'Tab' },
  { id: 'ctrl', label: 'ctrl', title: 'Control — applies to the next key' },
  { id: 'up', label: '↑', title: 'Up arrow' },
  { id: 'down', label: '↓', title: 'Down arrow' },
  { id: 'left', label: '←', title: 'Left arrow' },
  { id: 'right', label: '→', title: 'Right arrow' },
  { id: 'pipe', label: '|', title: 'Pipe' },
  { id: 'slash', label: '/', title: 'Slash' },
  { id: 'dash', label: '-', title: 'Hyphen' },
  { id: 'tilde', label: '~', title: 'Tilde' },
]

/* ----------------------------------------------------------------- state -- */

export interface KeyBarState {
  /** Ctrl is armed and folds into the next key pressed. */
  ctrl: boolean
}

export const KEY_BAR_IDLE: KeyBarState = { ctrl: false }

export interface KeyPress {
  /** Bytes for the session. Empty when the press only changed the modifier. */
  data: string
  state: KeyBarState
}

/* ------------------------------------------------------- control mapping -- */

/**
 * The ASCII control byte for a character, or null when there is not one.
 *
 * This is the actual rule rather than a table of guesses: a control character
 * is the printable one with bits 6 and 7 cleared, which is only defined for
 * `@` through `_` (0x40–0x5F) and, by the same masking, the lowercase letters.
 * Space is `@` with a different name, and `?` is the odd one out at 0x7F.
 *
 * Everything else returns null, deliberately. Terminals disagree about
 * Ctrl+`/` — the mask says 0x0F, several emulators send 0x1F — and this client
 * has no way to check which one is on the other end of the socket, so it does
 * not invent an answer. `|` is in the same position.
 */
export function controlByteForChar(char: string): string | null {
  if (char.length !== 1) return null
  if (char === ' ') return '\x00'
  if (char === '?') return '\x7f'
  const code = char.charCodeAt(0)
  if (code >= 0x40 && code <= 0x5f) return String.fromCharCode(code & 0x1f)
  if (code >= 0x61 && code <= 0x7a) return String.fromCharCode(code & 0x1f)
  return null
}

/**
 * The control byte for a `KeyboardEvent.code`, used only for hardware keyboards.
 *
 * `code` names the physical key. `key` and `keyCode` name what the character
 * layer decided the key produced, and on iOS with a hardware keyboard that
 * layer is wrong: through xterm.js 6.0.0, Safari reports Ctrl+C as
 * `keyCode: 13`. xterm's key handler reads `keyCode`, matches Enter, and sends
 * a carriage return — so the one keystroke whose entire job is to stop a
 * runaway process instead submits a blank line to it, and the process keeps
 * running. `code` still reads `KeyC`, because it never passes through the
 * layer that gets it wrong.
 *
 * So Ctrl chords are decoded here from `code` and written to the session
 * directly, and xterm is told not to handle the event.
 *
 * Driven rather than assumed: the malformed event was dispatched at a real
 * xterm 6.0.0 in this client — `{ key: 'c', code: 'KeyC', ctrlKey: true }` with
 * `keyCode` forced to 13 — and the frame that went out carried byte 0x03, not
 * 0x0d. What has *not* been reproduced here is the bug itself on real hardware;
 * there is no iOS device in this loop. That is why the workaround is written to
 * be harmless where the bug is absent: on a correct browser it produces exactly
 * the byte xterm would have produced anyway.
 */
export function controlByteForCode(code: string): string | null {
  const letter = /^Key([A-Z])$/.exec(code)
  if (letter) return controlByteForChar(letter[1].toLowerCase())
  // The four bracket-ish keys are worth having: Ctrl+[ is Escape on a keyboard
  // that has no Escape, which is every recent MacBook-style layout under iOS.
  if (code === 'BracketLeft') return '\x1b'
  if (code === 'Backslash') return '\x1c'
  if (code === 'BracketRight') return '\x1d'
  if (code === 'Space') return '\x00'
  return null
}

/* ---------------------------------------------------------------- presses -- */

/** Arrow keys, plain and with Ctrl (CSI 1;5 is the modifier terminals expect). */
const ARROWS: Record<'up' | 'down' | 'left' | 'right', { plain: string; ctrl: string }> = {
  up: { plain: '\x1b[A', ctrl: '\x1b[1;5A' },
  down: { plain: '\x1b[B', ctrl: '\x1b[1;5B' },
  right: { plain: '\x1b[C', ctrl: '\x1b[1;5C' },
  left: { plain: '\x1b[D', ctrl: '\x1b[1;5D' },
}

const LITERALS: Record<'pipe' | 'slash' | 'dash' | 'tilde', string> = {
  pipe: '|',
  slash: '/',
  dash: '-',
  tilde: '~',
}

/**
 * One key-bar tap: the bytes it produces and the state it leaves behind.
 *
 * Pure, and returns a new state rather than mutating one, so the caller can
 * render from the result and a test can assert a whole sequence.
 */
export function pressKeyBarKey(state: KeyBarState, id: KeyBarKeyId): KeyPress {
  if (id === 'ctrl') return { data: '', state: { ctrl: !state.ctrl } }

  const spent: KeyBarState = { ctrl: false }

  if (id === 'esc') {
    // Ctrl+[ *is* Escape, so an armed Ctrl changes nothing here rather than
    // producing a second, different escape.
    return { data: '\x1b', state: spent }
  }
  if (id === 'tab') {
    // Same argument: Ctrl+I is Tab.
    return { data: '\t', state: spent }
  }

  if (id === 'up' || id === 'down' || id === 'left' || id === 'right') {
    const arrow = ARROWS[id]
    return { data: state.ctrl ? arrow.ctrl : arrow.plain, state: spent }
  }

  const literal = LITERALS[id]
  // Ctrl over a character with no control byte sends the character alone. The
  // modifier is still spent — see the note at the top about the key after.
  return { data: (state.ctrl ? controlByteForChar(literal) : null) ?? literal, state: spent }
}

/**
 * A character typed on the soft keyboard, folded through an armed Ctrl.
 *
 * xterm hands us the character it decoded; when Ctrl is armed this is the only
 * place the chord can be formed, because the soft keyboard never reports a
 * modifier that a finger tapped on our own toolbar.
 */
export function pressCharacter(state: KeyBarState, char: string): KeyPress {
  if (!state.ctrl) return { data: char, state }
  return { data: controlByteForChar(char) ?? char, state: { ctrl: false } }
}

/* -------------------------------------------------------------------- DOM -- */

export interface KeyBarHandlers {
  /** Bytes to send to the session. Never called with an empty string. */
  onData(data: string): void
  /** Called whenever the armed state changes, so the caller can restyle. */
  onModifierChange?(state: KeyBarState): void
}

export interface KeyBarHandle {
  readonly element: HTMLElement
  /** Fold a character from the keyboard through the armed modifier. */
  handleCharacter(char: string): string
  state(): KeyBarState
  destroy(): void
}

/**
 * Build the row.
 *
 * `pointerdown` with `preventDefault`, not `click`: on iOS a `click` on a
 * button blurs the terminal's hidden textarea, which dismisses the soft
 * keyboard, and the keyboard reopening under every Ctrl+C makes the row
 * unusable. Preventing the default on pointerdown keeps focus where it is.
 */
export function createKeyBar(handlers: KeyBarHandlers): KeyBarHandle {
  const element = document.createElement('div')
  element.className = 'keybar'
  element.setAttribute('role', 'toolbar')
  element.setAttribute('aria-label', 'Terminal keys')

  let state: KeyBarState = KEY_BAR_IDLE
  const buttons = new Map<KeyBarKeyId, HTMLButtonElement>()

  const paint = (): void => {
    const ctrl = buttons.get('ctrl')
    if (ctrl) {
      ctrl.classList.toggle('is-armed', state.ctrl)
      ctrl.setAttribute('aria-pressed', String(state.ctrl))
    }
    handlers.onModifierChange?.(state)
  }

  const press = (id: KeyBarKeyId): void => {
    const result = pressKeyBarKey(state, id)
    state = result.state
    paint()
    if (result.data !== '') handlers.onData(result.data)
  }

  const onPointerDown = (event: PointerEvent): void => {
    const target = event.target
    if (!(target instanceof HTMLElement)) return
    const id = target.dataset.key as KeyBarKeyId | undefined
    if (!id) return
    event.preventDefault()
    press(id)
  }

  for (const key of KEY_BAR) {
    const button = document.createElement('button')
    button.type = 'button'
    button.className = `keybar__key keybar__key--${key.id}`
    button.dataset.key = key.id
    button.textContent = key.label
    button.title = key.title
    button.setAttribute('aria-label', key.title)
    if (key.id === 'ctrl') button.setAttribute('aria-pressed', 'false')
    buttons.set(key.id, button)
    element.appendChild(button)
  }

  element.addEventListener('pointerdown', onPointerDown)

  return {
    element,
    handleCharacter(char: string): string {
      const result = pressCharacter(state, char)
      if (result.state.ctrl !== state.ctrl) {
        state = result.state
        paint()
      }
      return result.data
    },
    state: () => state,
    destroy(): void {
      element.removeEventListener('pointerdown', onPointerDown)
      element.remove()
    },
  }
}
